### ISSUE TITLE
libmusl with static OpenSSL?

### ISSUE THREAD

**Issue by emk** [rust-embedded/cross#21]
This looks really cool!

There's a nice trick with musl-libc, where you can build a 100% static Rust executable that works on any distro with a reasonably modern kernel. It's also more-or-less possible to make this work with static OpenSSL.  For an example, see https://github.com/emk/rust-musl-builder.

Is there any support for doing something like this with cross? If so, I'd be happy to retire rust-musl-builder and point people at cross!

**Comment by japaric** [comment]
It ... already works? `cross build --target x86_64-unknown-linux-musl` on the Cargo repo gives me a statically linked Cargo.

If you find sometimes that doesn't work, let me know and I'll look into it!

**Comment by emk** [comment]
Sweet! I will look into this tomorrow and try converting a bunch of production projects at work, and let you know how it goes.

**Comment by jimmycuadra** [comment]
Follow up to this: Is there a reason OpenSSL is not included in the non-x86 musl targets? I'm interested in using cross to create static binaries targeting Raspberry Pi 3, but need OpenSSL. Previously I've used https://github.com/clux/muslrust for creating static binaries, but that project does not support multiple architectures. In clux/muslrust#11, we discuss this project superseding that one.

**Comment by japaric** [comment]
> Is there a reason OpenSSL is not included in the non-x86 musl targets? 

I couldn't get OpenSSL to compile using the MUSL toolchains last time I tried (and the alternative of compiling OpenSSL using the glibc toolchain resulted in link errors when trying to produce binaries, iirc). Sadly, I don't remember the details :-(.

**Comment by jimmycuadra** [comment]
@clux Any insight here?

**Comment by clux** [comment]
Hi.

No, not really. OpenSSL is difficult. In all my attempts at getting the linux musl build to work, I had to get OpenSSL compiled correctly with musl-gcc. Any attempt where I accidentally/purposefully used native gcc, I'd end up with link errors at the end of cargo build.

Also, I just gotta say that I really like how the docker images in here are set up. The tricks to compile then apt purge build dependencies for instance. Nice way of minimising dependencies and not accidentally relying on the wrong things (I ended up spending way too much time getting pkgconfig to do the right thing with custom prefixes).
